source/common/filesystem/posix/filesystem_impl.cc

```diff
@@ -130,12 +130,15 @@ bool InstanceImplPosix::illegalPath(const std::string& path) {
   // instances poking in bad places. We may have to consider conditioning on
   // platform in the future, growing these or relaxing some constraints (e.g.
   // there are valid reasons to go via /proc for file paths).
-  // TODO(htuch): Optimize this as a hash lookup if we grow any further.
-  if (absl::StartsWith(canonical_path.rc_, "/dev") ||
-      absl::StartsWith(canonical_path.rc_, "/sys") ||
-      absl::StartsWith(canonical_path.rc_, "/proc")) {
-    return true;
+  static const char* const paths[] = {"/dev", "/sys", "/proc"};
+  absl::string_view canonical(canonical_path.rc_);
+  const auto offset = canonical.find('/', 1);
+  for (const auto& path : paths) {
+    if (canonical.compare(0, offset, path) == 0) {
+      return true;
+    }
```

Comment:
What do you think of using `absl::StartsWith()` to detect paths that start with one of the banned prefixes, and then check if it's an exact match or if canonical has a '/' immediately following the prefix? 
Something like:
```
for (const auto& path : paths) {
  if (absl::StartsWith(canonical, path)) {
    if (canonical.size() == path.size() || canonical[path.size()] == '/') {
      return true;
    }
  }
}
```

This would require storing std::string instances in the static, which implies creating a static accessor method using the CONSTRUCT_ON_FIRST_USE macro to obtain the prefixes.

My motivation is that I think this is a little easier to follow and would allow additions of banned prefixes that have multi-part paths. That said, I don't feel super strongly about it.

Reply:
updated accordingly